src/V3List.h

```diff
@@ -140,4 +141,199 @@ class V3ListEnt final {
 
 //============================================================================
 
+// The list links, to be added as a member to 'Element'. They are considered mutable, even in a
+// 'const Element', as they are only used for storing 'Element' in a collection.
+template <typename Base>
+class V3List2Links final {
+    // The List itself, but nothing else can access the link pointers
+    template <typename B, V3List2Links<B> B::*P, typename>
+    friend class V3List2;
+
+    Base* m_nextp = nullptr;  // Next element in list
+    Base* m_prevp = nullptr;  // Previous element in list
+
+public:
+    V3List2Links() = default;
+    ~V3List2Links() {
+#ifdef VL_DEBUG
+        m_nextp = reinterpret_cast<Base*>(1);
+        m_prevp = reinterpret_cast<Base*>(1);
+#endif
+    }
+    VL_UNCOPYABLE(V3List2Links);
+    VL_UNMOVABLE(V3List2Links);
+};
+
+template <typename Base, V3List2Links<Base> Base::*LinksPointer, typename Element = Base>
+class V3List2 final {
+    static_assert(std::is_base_of<Base, Element>::value, "'Element' must be a subtype of 'Base");
+    Base* m_headp = nullptr;
+    Base* m_tailp = nullptr;
+
+    // Given the Element, return the Links. The links are always mutable, even in const elements.
+    VL_ATTR_ALWINLINE
+    static V3List2Links<Base>& toLinks(const Base& element) {
+        return const_cast<Base&>(element).*LinksPointer;
+    }
+
+    // Bare-bones iterator class for List. This is just enough to support range based for loops and
+    // basic usage. Feel free to extend as required.
+    template <typename IteratorElement>
+    class Iterator final {
```

Comment:
```suggestion
    class iterator final {
```

Although Iterator is correct Verilator naming, "::iterator" is the expected general standard.

Reply:
This is not that `iterator`. See below for `using iterator = Iterator<Element>` and `using const_iterator = Iterator<const Element>`, though I'm happy to change it to `IteratorImpl` of whatever.

Reply:
Ah, please pick a different name then, IteratorImpl works for me.